git.latest: add option to hard-reset only when there are remote changes

terminalmage opened the pull request:
This adds a third potential value for the `force_reset` argument, `'remote-changes'` which if used will not discard uncommitted changes unless the local clone is not up-to-date with the remote repository.

Resolves #35888.

terminalmage committed:
Add `remote-changes` value for git.latest state

This allows for local changes to be discarded only when the repo is not
already up-to-date.

terminalmage committed:
Add integration tests for new remote-changes support

rallytime commented on salt/modules/git.py:
```diff
@@ -2115,6 +2115,82 @@ def diff(cwd,
+            be observed (i.e. the encoding specified by the locale will be
+            tried first, and if that fails, UTF-8 will be used as a fallback).
+
+        .. versionadded:: 2018.3.1
```
Also probably not needed, I think. :)

rallytime reviewed (commented):
I just have a couple of really picky comments, otherwise this looks really nice and ++ for tests. :)

terminalmage committed:
Minor docstring fixes

This removes some versionadded/versionchanged lines that were
unnecessary in a brand new function, but ended up there because of
copypasta. It also updates the output_encoding documentation to match
the updated docs from a recent PR that hasn't yet been merged from
2018.3 to develop.

terminalmage commented:
@rallytime good catch, updated. Thanks!
